### PR TITLE
feat: main vit training logic

### PR DESCRIPTION
**Pull request by hlee9212**
This PR contains the main ViT training logic (as well as its default configuration file). It's written in **PyTorch Lightning**, which is a light wrapper around PyTorch - mostly to remove boilerplate code and sort of "force" you to modularize your code so you don't have one huge `train()` function. 😄 

The convention is to write a `YourClassLitModule(LightningModule)` class and override the following functions:
```python
def __init__(self):                     # Define model layers
def setup(self, stage):                # Stage-specific setup (fit/test/etc.)
def training_step(self, batch, idx):   # Logic for one training step
def validation_step(self, batch, idx): # Logic for one validation step
def test_step(self, batch, idx):       # Logic for one test step
def predict_step(self, batch, idx):    # Inference logic
def configure_optimizers(self):        # Optimizers and schedulers
```

Note: With Lightning, you don’t need to manually call loss.backward(), optimizer.step(), or optimizer.zero_grad() (these are the "boilerplate" that are removed). It should be pretty intuitive to read (relevant documentation page: https://lightning.ai/docs/pytorch/stable/common/lightning_module.html), but feel free to ask any questions!

### CODE REVIEW

File: pytorch/src/models/base_vit_module.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    def on_test_epoch_end(self) -> None:
+        """Reset test metrics at the end of each test epoch.
+
+        This method is called automatically by PyTorch Lightning after each test epoch to ensure
+        metrics are properly reset for the next epoch.
+        """
+        for metric in self.test_metrics.values():
+            metric.reset()
+
+    def predict_step(self, batch: BatchDict, batch_idx: int) -> PredictOutput:
```
Comment (nielsleadholm) [inline pytorch/src/models/base_vit_module.py:338]: **question:** Is `predict_step` used in some of the follow-up code you plan on merging? At the moment it looks like it's not used, in which case we could consider removing it.

File: pytorch/src/models/base_vit_module.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+            "predicted_quaternion": pred_quaternion,
+            "object_id": object_id,
+            "unit_quaternion": unit_quaternion,
+        }
+
+    def setup(self, stage: Optional[str]) -> None:
+        """Set up the model for the specified stage.
+
+        Args:
+            stage: The current stage ('fit', 'validate', 'test', or 'predict').
```
Comment (nielsleadholm) [inline pytorch/src/models/base_vit_module.py:378]: **question:** Is there a specific reason `fit` is used here, and `train` elsewhere? Just want to check it's not an error.

File: pytorch/src/models/base_vit_module.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                  Can be None during initialization.
+        """
+        if self.hparams.compile and stage == "fit":
+            self.net = torch.compile(self.net)
+
+    def configure_optimizers(self) -> OptimizerConfig:
```
Comment (nielsleadholm) [inline pytorch/src/models/base_vit_module.py:384]: **question:** This might also be clearer after a future PR, but where do the early-stopping number of epochs come in? Currently most values are hard-coded. I understand why you calculate warmup steps based on 200 (i.e. to not vary warmup steps across HP tuning and final training), but just want to double check I understand where e.g. 25 or 75 epochs eventually comes in.

**Review by nielsleadholm — COMMENTED** [review]
Overall looks really good, just a few clarifying questions.

**Review by nielsleadholm — APPROVED** [review]
